geode-core/src/main/java/org/apache/geode/management/internal/configuration/functions/GetClusterConfigurationFunction.java

```diff
@@ -32,20 +30,35 @@
 
   @Override
   public void execute(FunctionContext context) {
-    InternalClusterConfigurationService clusterConfigurationService =
-        InternalLocator.getLocator().getSharedConfiguration();
-
     Set<String> groups = (Set<String>) context.getArguments();
+    InternalLocator internalLocator = InternalLocator.getLocator();
+    logger.info("Received request for configuration: {}", groups);
 
-    logger.info("Received request for configuration  : {}", groups);
+    // Return exception to the caller so startup fails fast.
+    if (!internalLocator.isSharedConfigurationEnabled()) {
```

Comment:
should we check for isSharedConfigurationRunning() here also? If it's enabled but not in a running state yet, we should return this error too, right?

Reply:
Hello @jinmeiliao,

I don't think so, but I might be wrong here... the fact that the function is executed on a locator where the cluster configuration service is disabled (`internalLocator.isSharedConfigurationEnabled() == false`) can be considered as unexpected and that's the reason why the `IllegalStateException` is returned to the caller (`ClusterConfigurationLoader`).
It's possible (and expected if members are started up concurrently), however, for the cluster configuration service to be enabled but not yet running because it's actually starting up (`internalLocator.isSharedConfigurationEnabled() == true && internalLocator.isSharedConfigurationRunning() == false), in which case we shouldn't throw an exception but return `null` from the function. This allows the caller (`ClusterConfigurationLoader`) to decide whether to fail fast or keep retrying for some time, as it's currently doing [here](https://github.com/apache/geode/blob/d79282134c2064a09bfe418afcedaaf4c3741347/geode-core/src/main/java/org/apache/geode/internal/cache/ClusterConfigurationLoader.java#L278).
Please let me know your thoughts.
Cheers.

Reply:
Hey @jinmeiliao,

Thanks for the heads up, I haven't realized that we were already requesting the cluster configuration service *only* from locators where the service is actually enabled.
Either way, I have to disagree again... returning `null` when the cluster configuration service is not enabled can slow down the startup process altogether: the server will retry six times with a delay of 10 seconds after each retry, meaning that the startup failure will happen only after one minute. This shouldn't happen as we only request the cluster configuration from locators that have it enabled, but it's better to be on the safer side and fail fast on startup:

```
    int attempts = 6;
    OUTER: while (attempts > 0) {
      for (InternalDistributedMember locator : locatorList) {
        logger.info("Attempting to retrieve cluster configuration from {} - {} attempts remaining",
            locator.getName(), attempts);
        response = requestConfigurationFromOneLocator(locator, groups);
        if (response != null) {
          break OUTER;
        }
      }

      try {
        Thread.sleep(10000);
      } catch (InterruptedException ex) {
        break;
      }

      attempts--;
    }
```

Moreover, and thinking more about the `GetClusterConfigurationFunction` API, is more clear to return different things for different situations: if we return `null` on both scenarios (cluster configuration not enabled or cluster configuration enabled but not fully loaded yet) the user of the API (there's none at the moment other than `ClusterConfigurationLoader`, but it could be in the future...) doesn't get a clear perspective of what's happening... throwing an `IllegalStateException` when the cluster configuration service is not enabled and returning `null` when the service is enabled but not yet fully loaded is far more clear.
We could even refactor the `GetClusterConfigurationFunction` as a whole to return a typed response containing more accurate information about the status of the cluster configuration service without any exceptions at all, but that's another improvement, for the future maybe.
Please let me know what you think, I can make the changes if you still believe they are needed after reading the above.